src/FormControl/_FormControlContext.tsx

```diff
@@ -0,0 +1,59 @@
+import {createContext, useContext} from 'react'
+import {FormValidationStatus} from '../utils/types/FormValidationStatus'
+import {FormControlProps} from './FormControl'
+
+export interface FormControlContext extends Pick<FormControlProps, 'disabled' | 'id' | 'required'> {
+  captionId?: string
+  validationMessageId?: string
+  validationStatus?: FormValidationStatus
+}
+
+export const FormControlContext = createContext<FormControlContext | null>(null)
+
+interface FormControlForwardedProps extends Omit<FormControlContext, 'captionId' | 'validationMessageId'> {
+  ['aria-describedby']?: string
+}
+
+/**
+ * Make any component compatible with `FormControl`'s automatic wiring up of accessibility attributes & validation by
+ * reading the props from this hook and handling them / assigning them to the underlying form control. If used outside
+ * of `FormControl`, this hook has no effect.
+ *
+ * @param externalProps The external props passed to this component. If provided, these props will be merged with the
+ * `FormControl` props, with external props taking priority. This is also used for validating the external props,
+ * logging warnings to the console if there are conflicts.
+ */
+export function useFormControlForwardedProps<P extends FormControlForwardedProps>(externalProps: P): P {
+  const context = useContext(FormControlContext)
+  if (!context) return externalProps
+
+  if (externalProps.id) {
+    // eslint-disable-next-line no-console
+    console.warn(
+      `instead of passing the 'id' prop directly to the input component, it should be passed to the parent component, <FormControl>`,
+    )
+  }
+
+  if (externalProps.disabled) {
+    // eslint-disable-next-line no-console
+    console.warn(
+      `instead of passing the 'disabled' prop directly to the input component, it should be passed to the parent component, <FormControl>`,
+    )
+  }
+
+  if (externalProps.required) {
+    // eslint-disable-next-line no-console
+    console.warn(
+      `instead of passing the 'required' prop directly to the input component, it should be passed to the parent component, <FormControl>`,
+    )
+  }
```

Comment:
I wonder if these warnings should be suppressible (or just removed altogether). I kept them to be consistent with current behavior, but I do think there are valid extensibility cases where you might not want these warnings.

For example, a `DatePicker` might have a `TextInput` as its anchor. The `DatePicker` may want to control exactly how these forwarded props are handled, since a datepicker is a much more complicated component than just a text input. So the `DatePicker` would override the `TextInput` forwarded props, causing the warning to occur because the `TextInput` is still calling this hook.

One possible way to suppress this would be for the `DatePicker` to wrap its `TextInput` in a `FormControlContext.Provider value={null}`. But then we'd have to expose `FormControlContext` and I don't think we want to do that.

Reply:
I love the example that you brought up, I added a comment on our v37 release issue to try and capture this: https://github.com/github/primer/issues/1715#issuecomment-1682938944 as this is likely something that will stick around for v35 and v36.

Feel free to edit/add anything to the comment if you feel like there are other examples or situations that really demonstrate why this change is important 👀